### PR TITLE
Time integrators

### PR DESCRIPTION
**Pull request by sanguinariojoe**
**THIS PR DEPENDS ON https://github.com/FloatingArrayDesign/MoorDyn/pull/224**

Well, I am afraid this will require quite a bit of discussion on several of the changes I implemented:

* Fixed the Adams-Bashforth schemes [here](https://github.com/FloatingArrayDesign/MoorDyn/commit/5a03001c08f736d9a246dd5a7ed22d318d5f6a6e). I suppose this does not require any kind of discussion at all.
* Added a CFL ([Courant–Friedrichs–Lewy](https://en.wikipedia.org/wiki/Courant%E2%80%93Friedrichs%E2%80%93Lewy_condition)) option, in a such a way the user can either define the time step (dtM) or the CFL factor (cfl). He can even define both, so the most restrictive is considered.
* Removed the Z presevation check on the catenary solver (see [here](https://github.com/core-marine-dev/MoorDyn/commit/6570d0281036b060219cc45546a98398f94d610c)). To be honest, I do not remember if we added that, or it was always there. Anyway, if we want that back, we need to use the relative error instead. Otherwise, the errors for big depths will fall inside that condition (600 meters seems enough)
* Seabed is only considered now on the catenary solver when one of the line-ends is close enough to the bottom. This fix the initialization in many situations, and I do not think it is worsening the results on any context.
* Added Local-timestep versions of the Euler and the Adams-Bashforth time schemes. Tests on the lEuler have been created
* I substituted the previous initial condition by a stationary solver, which is essentially the same than before but with an infinitely large dissipation (indeed the trick is that velocity is completely removed). The main benefit is that the new solver is unconditionally stable (with the booster it might get unstable for a while though). Some other benefits are that it is able to converge to a way more precise solution. On the other hand, the null velocity might means that the solution would take significantly longer to be reached, specially when the catenary solution fails. Anyway, since the user still have the possibility of saving and loading, I do not see much benefit on the previous approach, which can still be used anyway.

### CODE REVIEW

**Comment by sanguinariojoe** [comment]
Hey guys! I uploaded a Newmark scheme. An Average Constant Acceleration (ACA) one to be more specific, with relaxation, which is able to remain stable with significantly larger time steps.

@mattEhall and @RyanDavies19 , it would be great if you can review this and give feedback.

**Comment by RyanDavies19** [comment]
@sanguinariojoe just keeping you in the loop, Matt and I have both been pretty busy with other work so we haven't gotten the chance to dig too much into this yet. It's still on our radar, and hopefully I'll be able to get to it soon. Currently I am working on adding vortex induced vibrations to MoorDyn.

**Comment by sanguinariojoe** [comment]
> Currently I am working on adding vortex induced vibrations to MoorDyn

Cool! I was told to avoid this on the past because apparently those are engineered out, so our digital twins shall not take them into account

**Comment by sanguinariojoe** [comment]
Hey @RyanDavies19 !

This branch is evolving quite much... I am indeed never using dev branch anymore.

- The stationary solution is way more powerful than the previous Drag increased Initial Condition generator. Probably there is room for even further improvements.
- The implicit schemes (specially the midpoint) are clearly outperforming the explicit ones.
- I added auto-mem-checks, which are really handy.
- I fixed several errors

I really think we should start addressing this before both branches start diverging too much

**Comment by RyanDavies19** [comment]
@sanguinariojoe agreed. I will be reviewing today now that I have some time available for MDC work. Thank you for including the docs, that is very helpful. My biggest concerns are making sure that 1) we retain backwards compatibility in the form of input files, and 2) the new input file is well documented (which it seems you've already done a lot of). Many of the questions I see in the OpenFAST issues/forum are from confusions between the two different code versions so keeping things clear will be important.

**Comment by sanguinariojoe** [comment]
Well, actually i want to document it a bit further. The discussion on the
relaxation factor is unfinished

**Comment by RyanDavies19** [comment]
@sanguinariojoe okay that would be good. I'll add comments in the code where some slight changes might be needed

File: docs/inputs.rst [hunk 1]
```diff
@@ -165,7 +165,8 @@ parenthesis). As such, they are all optional settings, although some of them (su
 size) often need to be set by the user for proper operation. The list of possible options (with any 
 default value provided in parentheses) is:
 
- - dtM (0.001) – desired mooring model time step (s)
+ - dtM (3.402823e+38) – desired mooring model maximum time step (s)
+ - CFL (0.5) – desired mooring model maximum CFL factor
```
Comment (RyanDavies19) [inline docs/inputs.rst:169]: These CFL notes are helpful, but need to be in the V2 section (this is V1). Additionally they should be under the line "The following MoorDyn-C options are not supported by MoorDyn-F:"

File: docs/tschemes.rst [hunk 2]
```diff
@@ -0,0 +1,314 @@
+Time Schemes
+============
+.. _tschemes:
+
+MoorDyn 2 is deployed with several time schemes, with different features,
```
Comment (RyanDavies19) [inline docs/tschemes.rst:5]: "MoorDyn-C version 2 is deployed..."

File: docs/inputs.rst (hunk 1 above)
Comment (sanguinariojoe) [reply docs/inputs.rst]: Fixed on https://github.com/FloatingArrayDesign/MoorDyn/pull/182/commits/5c3a404c3b2ee7809ebda94754e611c4c4f30691

File: docs/tschemes.rst (hunk 2 above)
Comment (sanguinariojoe) [reply docs/tschemes.rst]: Fixed on https://github.com/FloatingArrayDesign/MoorDyn/pull/182/commits/1bb95b347bb7c25d949dfde5ce6b5e448a2a073a

File: docs/tschemes.rst [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+strengths and weaknesses.
+
```
[the remark this answers is not in the document]
Comment (sanguinariojoe) [reply docs/tschemes.rst:12]: Fixed on https://github.com/FloatingArrayDesign/MoorDyn/pull/182/commits/e4b95e01ad0a8e090fb2e896e91054764f82f40c

**Comment by sanguinariojoe** [comment]
I rebased this PR on https://github.com/FloatingArrayDesign/MoorDyn/pull/224, so that PR shall be merged first

**Comment by sanguinariojoe** [comment]
Nice! I think #224 can be merged. Anyway the verifications against openfast
dev have to be made manually

**Comment by RyanDavies19** [comment]
@sanguinariojoe I agree it looks good to go. OpenFAST dev is a few weeks away from merging into main so we should be able to get the new reg tests in by then. If not it's a simple fix to point to the dev branch.

**Comment by RyanDavies19** [comment]
@sanguinariojoe Added a few more docs reviews. I will open a PR to your branch with the fixes. Overall though they look really good and are very helpful. Thanks for putting them together!

File: docs/inputs.rst [hunk 4]
```diff
@@ -548,16 +552,20 @@ indicate which version has that as the default. As such, they are all optional s
...
+   :ref:`driver file <MDF_driver_in>` dtC value or the OpenFAST time step.
+ - CFL (0.5) – Desired mooring model maximum Courant-Friedich-Lewy factor
+ - tScheme (RK2): The time integrator. It should be one of
+   Euler, LEuler, Heun, RK2, RK4, AB2, AB3, AB4, LAB2, LAB3, LAB4, 
+   BEuler\ *N*, Midpoint\ *N*, ACA\ *N*, Wilson\ *N*. Look at the
+   :ref:`time schemes documentation <tschemes>` to learn more about this.
  - g (9.81): The gravity acceleration (m/s^2)
  - rho (1025): The water density (kg/m^3)
- - WtrDpth (0.0): The water depth (m). In MoorDyn-F the bathymetry file path can be inputted here.
+ - WtrDpth (0.0): The water depth (m). In MoorDyn-C the bathymetry file path can be inputted here.
```
[the remark this answers is not in the document]
Comment (RyanDavies19) [reply docs/inputs.rst]: Resolved

File: docs/inputs.rst [hunk 5]
```diff
@@ -179,6 +179,11 @@ default value provided in parentheses) is:
  - ThreshIC (0.001) – convergence threshold for IC generation, acceptable relative difference 
    between three successive fairlead tension measurements (-)
 
+The internal time step is first taken from the dtM option, and then adjusted
```
[the remark this answers is not in the document]
Comment (RyanDavies19) [reply docs/inputs.rst]: Resolved

File: source/MoorDyn2.cpp [hunk 6]
```diff
@@ -87,7 +86,8 @@ moordyn::MoorDyn::MoorDyn(const char* infilename, int log_level)
   , ICTmax(120.0)
   , ICthresh(0.001)
   , WaveKinTemp(waves::WAVES_NONE)
-  , dtM0(0.001)
+  , dtM0((std::numeric_limits<real>::max)())
+  , cfl(0.5)
   , dtOut(0.0)
```
[the remark this answers is not in the document]
Comment (RyanDavies19) [reply source/MoorDyn2.cpp:91]: Resolved

**Review by RyanDavies19 — APPROVED** [review]

**Comment by RyanDavies19** [comment]
See https://github.com/OpenFAST/r-test/issues/127 for updated reg tests

File: source/MoorDyn2.cpp [hunk 7]
```diff
@@ -302,8 +297,30 @@ moordyn::MoorDyn::Init(const double* x, const double* xd, bool skip_ic)
 		ix += 3;
 	}
 
+	// Compute the timestep
+	for (auto obj : LineList)
+		dtM0 = (std::min)(dtM0, obj->cfl2dt(cfl));
+	for (auto obj : PointList)
```
[the remark this answers is not in the document]
Comment (sanguinariojoe) [reply source/MoorDyn2.cpp:524]: It is indeed. But in the future that might change so they might have their own requirements for stability or accuracy. A formal analysis shall be made

**Comment by sanguinariojoe** [comment]
> See [OpenFAST/r-test#127](https://github.com/OpenFAST/r-test/pull/127) for updated reg tests

Hey @RyanDavies19 , the new tests do not work to me... In OpenFAST I mean. It seems that all the drivers named like `BodiesAndRods_driver.dvr` shall be replaced by `md_driver.inp`

**Comment by RyanDavies19** [comment]
@sanguinariojoe Agreed lets merge. I talked with Matt and that's fine. Apologies for confusion with https://github.com/OpenFAST/r-test/pull/127, I should have opened that as a draft. It is a work in progress.

**Comment by sanguinariojoe** [comment]
Hurray! Thanks guys!

**Pull request merged**